File: backend/agents/agent_service.py
"""Main service class that handles all AI agent interactions."""

import json
from datetime import datetime
from typing import Any, Dict, List, Optional
import google.generativeai as genai

from .agent_types import (
    SafetyStatus,
    LearningState,
    AgentClassifierInput,
    AgentClassifierOutput,
    SafetyAgentInput,
    SafetyAgentOutput,
    QuestionAgentInput,
    QuestionAgentOutput,
    AnswerEvalAgentInput,
    AnswerEvalAgentOutput,
    InteractiveAgentInput,
    InteractiveAgentOutput,
    SummaryConsolidationAgentInput,
    SummaryConsolidationAgentOutput,
    ExplorationAgentOutput,
    ExplorationAgentInput,
    DeepDiveAgentInput,
    DeepDiveAgentOutput,
    FlashcardAgentInput,
    FlashcardAgentOutput,
    CheatsheetAgentInput,
    CheatsheetAgentOutput,
    MermaidAgentInput,
    MermaidAgentOutput,
    ConfigAgentInput,
    ConfigAgentOutput
)

from .implementations import (
    handle_exploration,
    handle_interactive,
    handle_question,
    handle_answer_eval,
    handle_classification,
    handle_safety,
    handle_summary,
    handle_deep_dive,
    handle_flashcard,
    handle_cheatsheet,
    handle_mermaid,
    handle_config
)

class AgentService:
    """Service class that manages all AI agent interactions."""

    def __init__(self, api_key: str):
        """Initialize the agent service with API key."""
        genai.configure(api_key=api_key)
        self.model = genai.GenerativeModel('gemini-pro')
        self.learning_state = self._initialize_learning_state()

    def _initialize_learning_state(self) -> LearningState:
        """Initialize a new learning state."""
        return LearningState(
            current_topic="",
            active_subtopic="",
            learning_path=[],
            progress={
                "completed_subtopics": [],
                "mastered_concepts": [],
                "needs_review": []
            },
            session_history=[]
        )

    def _add_to_session_history(self, entry: Dict[str, Any]) -> None:
        """Add an entry to the session history."""
        if not isinstance(entry, dict) or 'type' not in entry or 'content' not in entry:
            raise ValueError("Invalid session history entry format")
        
        entry['timestamp'] = datetime.now().isoformat()
        self.learning_state.session_history.append(entry)

    def _handle_answer_evaluation(self, topic: str) -> None:
        self.learning_state.awaiting_answer = False
        answer_eval_input = AnswerEvalAgentInput("", self.learning_state.active_subtopic, self.learning_state.current_topic, self.learning_state.last_question, topic)
        answer_eval = handle_answer_eval(self.model, answer_eval_input, self._call_agent)
        return AnswerEvalAgentOutput(
            is_correct=answer_eval.is_correct,
            feedback=answer_eval.feedback
        )

    def _call_agent(self, instructions: str, input_data: Any) -> Any:
        """Handle communication with the AI model."""
        print('\n=== Agent Call ===')
        print('Instructions:', instructions.split('\n')[0])
        print('Input:', json.dumps(getattr(input_data, "to_dict", lambda: input_data)(), indent=2))

        try:
            chat = self.model.start_chat(history=[
                {
                    'role': 'user',
                    'parts': [{'text': instructions}]
                },
                {
                    'role': 'model',
                    'parts': [{'text': 'I understand my role and instructions. Ready to process input.'}]
                }
            ])

            result = chat.send_message(json.dumps({
                **(input_data if isinstance(input_data, dict) else input_data.to_dict()),
                'response_format': 'json',
                'format_instructions': 'Return only valid JSON without any markdown formatting or additional text.'
            }))

            response = result.text
            print('Raw response:', response)

            try:
                json_start = response.find('{')
                json_end = response.rfind('}') + 1
                if json_start >= 0 and json_end > json_start:
                    clean_response = response[json_start:json_end]
                    parsed_response = json.loads(clean_response, strict=False)
                else:
                    return {
                        'status': SafetyStatus.SAFE,
                        'explanation': response.strip(),
                        'subtopics': [],
                        'prerequisites': [],
                        'summary': ''
                    }

                if not isinstance(input_data, SafetyAgentInput):
                    response_text = json.dumps(parsed_response, ensure_ascii=False).lower()
                    moderation_phrases = [
                        'cannot help',
                        'inappropriate',
                        'harmful',
                        'unacceptable',
                        "i'm sorry",
                        'i am sorry',
                        'i apologize',
                        'not appropriate',
                        'racism'
                    ]
                    
                    if any(phrase in response_text for phrase in moderation_phrases):
                        return {
                            'status': SafetyStatus.INAPPROPRIATE,
                            'explanation': "I apologize, but I cannot generate that type of content. Let's focus on something else."
                        }

                return parsed_response

            except json.JSONDecodeError as e:
                print(f'Error parsing JSON response: {e}')
                return {
                    'status': SafetyStatus.SAFE,
                    'explanation': response.strip(),
                    'subtopics': [],
                    'prerequisites': [],
                    'summary': ''
                }

        except Exception as e:
            print(f'Error in agent call: {e}')
            if 'SAFETY' in str(e):
                return {
                    'status': SafetyStatus.INAPPROPRIATE,
                    'explanation': "I apologize, but I cannot generate that type of content. Let's focus on something else."
                }
            return {
                'status': SafetyStatus.SAFE,
                'explanation': 'I encountered an error processing your request. Could you please rephrase it?',
                'subtopics': [],
                'prerequisites': [],
                'summary': ''
            }

    def run_safety_check(self, input_text: str) -> SafetyAgentOutput:
        """Run a safety check on user input."""
        print('\n=== Running Safety Check ===')
        print('Input:', input_text)
        print('Session history length:', len(self.learning_state.session_history))

        safety_input = SafetyAgentInput(
            user_input=input_text,
            latest_context_summary='\n'.join(
                entry['content'] for entry in self.learning_state.session_history
            )
        )

        return handle_safety(self.model, safety_input, self._call_agent)

def start_new_topic(self, topic: str, user_background: Optional[str] = None, current_topic: Optional[str] = None, active_subtopic: Optional[str] = None, session_history: Optional[List[str]] = None) -> ExplorationAgentOutput:
    """Begin a new learning topic."""
    print('\n=== Starting Agent Pipeline ===')
    print('Input:', topic)
    
    self.learning_state.current_topic = current_topic if current_topic is not None else topic
    self.learning_state.active_subtopic = active_subtopic if active_subtopic is not None else topic
    self.learning_state.session_history = session_history if session_history is not None else []

    safety_check = self.run_safety_check(topic)
    if safety_check.status != SafetyStatus.SAFE:
        return ExplorationAgentOutput(
            status=safety_check.status,
            explanation=safety_check.explanation,
            subtopics=[],
            prerequisites=[],
            summary=safety_check.explanation
        )

    classifier_input = AgentClassifierInput(
        user_input=topic,
        available_agents=[
            {'name': 'exploration', 'description': 'Explores new topics'},
            {'name': 'interactive', 'description': 'Handles questions and answers'},
            {'name': 'question', 'description': 'Generates quiz questions'},
            {'name': 'answerEval', 'description': 'Evaluates answers to questions'},
            {'name': 'deepDive', 'description': 'Provides detailed concept breakdowns'},
            {'name': 'flashcard', 'description': 'Creates study flashcards'},
            {'name': 'cheatsheet', 'description': 'Generates quick reference guides'},
            {'name': 'mermaid', 'description': 'Creates visual diagrams'},
            {'name': 'config', 'description': 'Handles system configuration'}
        ],
        latest_context_summary='\n'.join(
            entry['content'] for entry in self.learning_state.session_history
        )
    )

    classification = handle_classification(self.model, classifier_input, self._call_agent)

    if self.learning_state.awaiting_answer and self.learning_state.last_question:
        return self._handle_answer_evaluation(topic)

    context_summary = '\n'.join(
        entry['content'] for entry in self.learning_state.session_history
    )

    print("Agent: ", classification.next_agent)
    agent = classification.next_agent

    if agent == 'exploration':
        input_data = ExplorationAgentInput(
            user_prompt=topic,
            latest_context_summary=context_summary
        )
        return handle_exploration(self.model, input_data, self._call_agent)

    elif agent == 'interactive':
        input_data = InteractiveAgentInput(
            user_input=topic,
            latest_context_summary=context_summary
        )
        response = handle_interactive(self.model, input_data, self._call_agent)
        return ExplorationAgentOutput(
            status=SafetyStatus.SAFE,
            explanation=response.response,
            subtopics=[],
            prerequisites=[],
            summary=response.response
        )

    elif agent == 'question':
        input_data = QuestionAgentInput(
            subtopic=self.learning_state.active_subtopic,
            broader_topic=self.learning_state.current_topic,
            latest_context_summary=context_summary
        )
        response = handle_question(self.model, input_data, self._call_agent)
        self.learning_state.last_question = response.question
        self.learning_state.last_question_type = response.type
        self.learning_state.awaiting_answer = True
        return ExplorationAgentOutput(
            status=SafetyStatus.SAFE,
            explanation=response.question,
            subtopics=response.options if response.type == 'MCQ' else [],
            prerequisites=[],
            summary=response.question
        )

    elif agent == 'deepDive':
        input_data = DeepDiveAgentInput(
            subtopic=self.learning_state.active_subtopic,
            broader_topic=self.learning_state.current_topic,
            latest_context_summary=context_summary
        )
        response = handle_deep_dive(self.model, input_data, self._call_agent)
        return ExplorationAgentOutput(
            status=SafetyStatus.SAFE,
            explanation=response.breakdown,
            subtopics=[],
            prerequisites=[],
            summary=response.breakdown
        )

    elif agent == 'flashcard':
        input_data = FlashcardAgentInput(
            broader_topic=self.learning_state.current_topic,
            subtopic=self.learning_state.active_subtopic,
            latest_context_summary=context_summary
        )
        response = handle_flashcard(self.model, input_data, self._call_agent)
        return ExplorationAgentOutput(
            status=SafetyStatus.SAFE,
            explanation="Here are your study flashcards\n\n" + response.csv_content,
            subtopics=[],
            prerequisites=[],
            summary=context_summary
        )

    elif agent == 'cheatsheet':
        input_data = CheatsheetAgentInput(
            broader_topic=self.learning_state.current_topic,
            subtopic=self.learning_state.active_subtopic,
            latest_context_summary=context_summary
        )
        response = handle_cheatsheet(self.model, input_data, self._call_agent)
        return ExplorationAgentOutput(
            status=SafetyStatus.SAFE,
            explanation=response.content,
            subtopics=[],
            prerequisites=[],
            summary=response.content
        )

    elif agent == 'mermaid':
        input_data = MermaidAgentInput(
            broader_topic=self.learning_state.current_topic,
            subtopic=self.learning_state.active_subtopic,
            available_diagram_types=["graph", "flowchart", "sequence", "class", "state"],
            latest_context_summary=context_summary
        )
        response = handle_mermaid(self.model, input_data, self._call_agent)
        return ExplorationAgentOutput(
            status=SafetyStatus.SAFE,
            explanation=response.mermaid_code,
            subtopics=[],
            prerequisites=[],
            summary=context_summary
        )

    elif agent == 'config':
        input_data = ConfigAgentInput(
            user_input=topic,
            latest_context_summary=context_summary
        )
        response = handle_config(self.model, input_data, self._call_agent)
        return ExplorationAgentOutput(
            status=SafetyStatus.SAFE,
            explanation=response.prompt_addition,
            subtopics=[],
            prerequisites=[],
            summary=response.prompt_addition
        )

    else:
        input_data = ExplorationAgentInput(
            user_prompt=topic,
            latest_context_summary=context_summary
        )
        return handle_exploration(self.model, input_data, self._call_agent)


    def get_session_summary(self) -> SummaryConsolidationAgentOutput:
        """Generate a summary of the learning session."""
        input_data = SummaryConsolidationAgentInput(
            latest_context_summary='\n'.join(
                entry['content'] for entry in self.learning_state.session_history
            ),
            last_agent_input=None,
            last_agent_output=None
        )

        return handle_summary(self.model, input_data, self._call_agent) 

File: app/services/agentService.ts
import { GoogleGenerativeAI } from '@google/generative-ai';
import {
  EXPLORATION_AGENT_INSTRUCTIONS,
  DEEP_DIVE_AGENT_INSTRUCTIONS,
  INTERACTIVE_AGENT_INSTRUCTIONS,
  QUESTION_AGENT_INSTRUCTIONS,
  ANSWER_EVAL_AGENT_INSTRUCTIONS,
  AGENT_CLASSIFIER_INSTRUCTIONS,
  CONFIG_AGENT_INSTRUCTIONS,
  SAFETY_AGENT_INSTRUCTIONS,
  FLASHCARD_AGENT_INSTRUCTIONS,
  CHEATSHEET_AGENT_INSTRUCTIONS,
  MERMAID_AGENT_INSTRUCTIONS,
  SUMMARY_CONSOLIDATION_AGENT_INSTRUCTIONS
} from '../data/agents';
import { ApiConfig, validateApiKeys } from '../config/apiConfig';

import { getLocationBasedResources } from '../data/crisisResources';
import { SafetyStatus } from '../types/newAgents';

import type {
  ExplorationAgentInput,
  ExplorationAgentOutput,
  DeepDiveAgentInput,
  DeepDiveAgentOutput,
  InteractiveAgentInput,
  InteractiveAgentOutput,
  QuestionAgentInput,
  QuestionAgentOutput,
  AnswerEvalAgentInput,
  AnswerEvalAgentOutput,
  AgentClassifierInput,
  AgentClassifierOutput,
  ConfigAgentInput,
  ConfigAgentOutput,
  SafetyAgentInput,
  SafetyAgentOutput,
  FlashcardAgentInput,
  FlashcardAgentOutput,
  CheatsheetAgentInput,
  CheatsheetAgentOutput,
  MermaidAgentInput,
  MermaidAgentOutput,
  SummaryConsolidationAgentInput,
  SummaryConsolidationAgentOutput
} from '../types/newAgents';

interface LearningState {
  currentTopic: string;
  activeSubtopic: string;
  learningPath: string[];
  progress: {
    completedSubtopics: string[];
    masteredConcepts: string[];
    needsReview: string[];
  };
  sessionHistory: SessionHistoryEntry[];
  difficulty: string;
  lastQuizScore?: number;
  lastQuizAnswer?: string;
  lastQuestion?: string;  // Track the last question asked
  lastQuestionType?: 'quiz' | 'interactive';  // Track the type of question
  awaitingAnswer?: boolean;  // Flag to indicate we're waiting for an answer
}

interface SessionHistoryEntry {
  timestamp: string;
  type: 'explanation' | 'quiz' | 'feedback' | 'summary';
  content: string;
  outcome?: string;
}

// Main service class that handles all AI agent interactions
export class AgentService {
  private genAI: GoogleGenerativeAI;
  private model: any;
  private learningState: LearningState;

  constructor(apiKey?: string) {
    const key = apiKey || ApiConfig.GEMINI_API_KEY;
    
    // Validate API key
    if (!key) {
      console.error('Missing Gemini API Key - Learning Assistant will not function correctly');
    } else {
      console.log('Initializing AgentService with Gemini API');
    }
    
    this.genAI = new GoogleGenerativeAI(key);
    this.model = this.genAI.getGenerativeModel({ model: 'gemini-pro' });
    this.learningState = this.initializeLearningState();
    
    // Run API key validation to log any missing keys
    validateApiKeys();
  }

  // Sets up initial learning state with default values
  private initializeLearningState(): LearningState {
    return {
      currentTopic: '',
      activeSubtopic: '',
      learningPath: [],
      progress: {
        completedSubtopics: [],
        masteredConcepts: [],
        needsReview: []
      },
      sessionHistory: [],
      difficulty: 'beginner'
    };
  }

  // Handles communication with the AI model
  private async callAgent(instructions: string, input: any): Promise<any> {
    console.log('\n=== Agent Call ===');
    console.log('Agent type:', instructions.split('\n')[0]);
    console.log('Input:', JSON.stringify(input, null, 2));

    try {
      const chat = this.model.startChat({
        history: [
          {
            role: 'user',
            parts: [{ text: instructions }]
          },
          {
            role: 'model',
            parts: [{ text: 'I understand my role and instructions. Ready to process input.' }]
          }
        ]
      });

      let result;
      try {
        result = await chat.sendMessage(JSON.stringify({
          ...input,
          responseFormat: 'json',
          formatInstructions: 'Return only valid JSON without any markdown formatting or additional text.'
        }));
      } catch (sendError: any) {
        console.error('Send message error:', sendError);
        // Handle safety blocks and other send errors
        if (sendError.message?.includes('SAFETY')) {
          console.log('Response blocked by safety filters');
          return {
            status: SafetyStatus.INAPPROPRIATE,
            explanation: "I apologize, but I cannot generate that type of content. Let's focus on something else.",
            subtopics: [],
            prerequisites: [],
            summary: ""
          };
        }
        throw sendError;
      }

      const response = await result.response;
      const text = response.text();

      console.log('Raw response:', text);

      // Common phrases that indicate a moderation response
      const moderationPhrases = [
        'cannot help',
        'inappropriate',
        'harmful',
        'unacceptable',
        'i\'m sorry',
        'i am sorry',
        'i apologize',
        'not appropriate',
        'racism'
      ];
      
      // First check if it's a plain text response starting with these phrases
      const lowerText = text.toLowerCase();
      if (moderationPhrases.some(phrase => lowerText.startsWith(phrase))) {
        console.log('Detected moderation response from text start');
        return {
          status: SafetyStatus.INAPPROPRIATE,
          explanation: text.trim(),
          subtopics: [],
          prerequisites: [],
          summary: ""
        };
      }
      
      // Clean the response text
      const cleanedText = text
        .replace(/```json\n?|\n?```/g, '')  // Remove code blocks
        .replace(/^[^{]*({.*})[^}]*$/, '$1') // Extract JSON object
        .trim();
      
      console.log('Cleaned response:', cleanedText);
      
      // If the cleaned text starts with "I" or other non-JSON characters, treat it as a moderation response
      if (cleanedText.startsWith('I ') || cleanedText.startsWith('I\'m ') || cleanedText.startsWith('Im ')) {
        console.log('Detected non-JSON response starting with "I"');
        return {
          status: SafetyStatus.INAPPROPRIATE,
          explanation: cleanedText,
          subtopics: [],
          prerequisites: [],
          summary: ""
        };
      }
      
      try {
        console.log('Attempting to parse response as JSON...');
        const parsedResponse = JSON.parse(cleanedText);
        console.log('Successfully parsed JSON response');
        
        // Check if it's an error response
        if (parsedResponse.error) {
          console.log('Detected error in parsed response:', parsedResponse.error);
          return {
            status: SafetyStatus.INAPPROPRIATE,
            explanation: parsedResponse.error,
            subtopics: [],
            prerequisites: [],
            summary: ""
          };
        }
        
        // Check if any field contains moderation phrases
        const responseStr = JSON.stringify(parsedResponse).toLowerCase();
        if (moderationPhrases.some(phrase => responseStr.includes(phrase))) {
          console.log('Detected moderation phrase in parsed response');
          const message = parsedResponse.error || parsedResponse.message || parsedResponse.explanation || 
                         Object.values(parsedResponse).find(v => typeof v === 'string' && moderationPhrases.some(p => v.toLowerCase().includes(p))) ||
                         "I cannot help with that request.";
          return {
            status: SafetyStatus.INAPPROPRIATE,
            explanation: message,
            subtopics: [],
            prerequisites: [],
            summary: ""
          };
        }
        
        // For exploration agent, ensure required fields exist
        if (instructions === EXPLORATION_AGENT_INSTRUCTIONS) {
          console.log('Processing exploration agent response');
          return {
            ...parsedResponse,
            subtopics: parsedResponse.subtopics || ['Basic Overview'],
            prerequisites: parsedResponse.prerequisites || [],
            summary: parsedResponse.summary || "Let's explore this topic.",
            status: SafetyStatus.SAFE
          };
        }

        // For question agent, ensure required fields exist
        if (instructions === QUESTION_AGENT_INSTRUCTIONS) {
          console.log('Processing question agent response');
          return {
            ...parsedResponse,
            question: parsedResponse.question || "What do you understand about this topic?",
            type: parsedResponse.type || "MCQ",
            options: parsedResponse.options || ["I understand it well", "I need more explanation"],
            correctAnswer: parsedResponse.correctAnswer || parsedResponse.options?.[0] || "I understand it well"
          };
        }
        
        console.log('Returning parsed response');
        return parsedResponse;
      } catch (parseError) {
        console.error('JSON parse error:', parseError);
        console.log('Failed to parse text:', cleanedText);
        
        // If it looks like a moderation response, return it as such
        if (moderationPhrases.some(phrase => cleanedText.toLowerCase().includes(phrase))) {
          console.log('Detected moderation phrase in unparseable response');
          return {
            status: SafetyStatus.INAPPROPRIATE,
            explanation: cleanedText || "I cannot help with that request.",
            subtopics: [],
            prerequisites: [],
            summary: ""
          };
        }
        
        // Default error response
        console.log('Returning default error response');
        return {
          status: SafetyStatus.NEEDS_HELP,
          explanation: "I encountered an error processing your request. Let me help you with something else.",
          subtopics: ['Basic Overview'],
          prerequisites: [],
          summary: "Let's start with the basics."
        };
      }
    } catch (error: any) {
      console.error('\n=== Agent Call Error ===');
      console.error('Error:', error);
      
      // Check if it's a safety-related error
      if (error.message?.includes('SAFETY')) {
        console.log('Response blocked by safety filters');
        return {
          status: SafetyStatus.INAPPROPRIATE,
          explanation: "I apologize, but I cannot generate that type of content. Let's focus on something else.",
          subtopics: [],
          prerequisites: [],
          summary: ""
        };
      }
      
      return {
        status: SafetyStatus.NEEDS_HELP,
        explanation: "I encountered an error. How can I help you with something else?",
        subtopics: ['Basic Overview'],
        prerequisites: [],
        summary: "Let's start with the basics."
      };
    }
  }

  // Adds a new entry to session history with timestamp
  private addToSessionHistory(entry: SessionHistoryEntry) {
    console.log('\n=== Adding to Session History ===');
    console.log('Entry type:', entry.type);
    console.log('Content:', entry.content);
    this.learningState.sessionHistory.push(entry);
  }

  // Safety check that must be run before any other agent
  private async runSafetyCheck(input: string): Promise<SafetyAgentOutput> {
    console.log('\n=== Running Safety Check ===');
    console.log('Input:', input);
    console.log('Session history length:', this.learningState.sessionHistory.length);

    const safetyInput: SafetyAgentInput = {
      userInput: input,
      latestContextSummary: this.learningState.sessionHistory
        .map(h => h.content)
        .join('\n')
    };

    const safetyCheck = await this.callAgent(
      SAFETY_AGENT_INSTRUCTIONS,
      safetyInput
    ) as SafetyAgentOutput;

    console.log('Safety check status:', safetyCheck.status);

    if (safetyCheck.status !== SafetyStatus.SAFE) {
      try {
        console.log('Getting location-based resources...');
        // Get user's IP address
        const ipResponse = await fetch('https://api.ipify.org?format=json');
        const ipData = await ipResponse.json();
        const userIp = ipData.ip;
        console.log('User IP:', userIp);

        // Get location-based resources
        const resource = await getLocationBasedResources(userIp);
        console.log('Got resources for location');
        
        // Format resources with markdown
        const resourceDetails = [
          resource.description ? `**${resource.description}**` : '',
          resource.phone.length ? `📞 **Phone**: ${resource.phone.join(', ')}` : '',
          resource.website ? `🌐 **Website**: [Click here](${resource.website})` : '',
          resource.email ? `✉️ **Email**: [${resource.email}](mailto:${resource.email})` : ''
        ].filter(Boolean);

        // Add a more empathetic and structured response
        safetyCheck.explanation = `I understand you're going through a difficult time. Your life has value, and there are people who want to help.\n\n### Immediate Support Available\n\n${resourceDetails.join('\n\n')}\n\n**You're not alone.** These services are:\n- Free and confidential\n- Available 24/7\n- Staffed by caring professionals\n- Here to listen without judgment\n\nPlease reach out - taking that first step can make all the difference.`;
      } catch (error) {
        console.error('Error getting location-based resources:', error);
        // Fallback to default resources if location detection fails
        console.log('Using default resources');
        const resource = await getLocationBasedResources();
        safetyCheck.explanation = `I care about your wellbeing. Please reach out for support:\n\n- **${resource.description}**\n- 📞 **Phone**: ${resource.phone.join(', ')}\n- 🌐 **Website**: [Click here](${resource.website})\n\nYou're not alone. These services are available 24/7 and ready to help.`;
      }
    }

    return safetyCheck;
  }

  // Begins a new learning topic
  async startNewTopic(topic: string, userBackground?: string): Promise<ExplorationAgentOutput> {
    console.log('\n=== Starting Agent Pipeline ===');
    console.log('Input:', topic);
    
    // Run safety check first
    console.log('Running safety check...');
    const safetyCheck = await this.runSafetyCheck(topic);
    
    if (safetyCheck.status !== SafetyStatus.SAFE) {
      console.log('Safety check failed:', safetyCheck.status);
      return {
        status: safetyCheck.status,
        explanation: safetyCheck.explanation,
        subtopics: [],
        prerequisites: [],
        summary: ""
      };
    }

    // Use agent classifier to determine the appropriate agent
    console.log('Running agent classifier...');
    const classifierInput: AgentClassifierInput = {
      userInput: topic,
      availableAgents: [
        { name: 'exploration', description: 'Explores new topics' },
        { name: 'interactive', description: 'Handles questions and answers' },
        { name: 'question', description: 'Generates quiz questions' },
        { name: 'answerEval', description: 'Evaluates answers to questions' }
      ],
      latestContextSummary: this.learningState.sessionHistory
        .map(h => h.content)
        .join('\n')
    };

    const classification = await this.callAgent(
      AGENT_CLASSIFIER_INSTRUCTIONS,
      classifierInput
    ) as AgentClassifierOutput;
    console.log('Agent classification:', classification.nextAgent);

    // If we're awaiting an answer, handle it with answer evaluation
    if (this.learningState.awaitingAnswer && this.learningState.lastQuestion) {
      console.log('Processing answer to previous question...');
      const evalInput: AnswerEvalAgentInput = {
        subtopic: this.learningState.activeSubtopic,
        broaderTopic: this.learningState.currentTopic,
        questionAsked: this.learningState.lastQuestion,
        userQuestionAnswer: topic,
        latestContextSummary: this.learningState.sessionHistory
          .map(h => h.content)
          .join('\n')
      };

      const feedback = await this.callAgent(
        ANSWER_EVAL_AGENT_INSTRUCTIONS,
        evalInput
      ) as AnswerEvalAgentOutput;

      // Reset question state
      this.learningState.lastQuestion = undefined;
      this.learningState.awaitingAnswer = false;

      // Add to session history
      this.addToSessionHistory({
        type: 'feedback',
        content: feedback.feedback,
        timestamp: new Date().toISOString()
      });

      return {
        status: SafetyStatus.SAFE,
        explanation: feedback.feedback,
        subtopics: [],
        prerequisites: [],
        summary: ""
      };
    }

    // Handle based on classification
    switch (classification.nextAgent) {
      case 'question':
        console.log('Generating quiz question...');
        const quiz = await this.getQuizQuestion(this.learningState.activeSubtopic);
        this.learningState.lastQuestion = quiz.question;
        this.learningState.lastQuestionType = 'quiz';
        this.learningState.awaitingAnswer = true;
        
        // Add to session history
        this.addToSessionHistory({
          type: 'quiz',
          content: quiz.question,
          timestamp: new Date().toISOString()
        });

        return {
          status: SafetyStatus.SAFE,
          explanation: quiz.question,
          subtopics: quiz.type === 'MCQ' ? quiz.options || [] : [],
          prerequisites: [],
          summary: ""
        };

      case 'answerEval':
        console.log('Evaluating answer...');
        if (!this.learningState.lastQuestion) {
          return {
            status: SafetyStatus.SAFE,
            explanation: "I don't see a previous question to evaluate. What would you like to learn about?",
            subtopics: [],
            prerequisites: [],
            summary: ""
          };
        }
        // Handle answer evaluation (already covered in the earlier check)
        return {
          status: SafetyStatus.SAFE,
          explanation: "Let me evaluate your answer...",
          subtopics: [],
          prerequisites: [],
          summary: ""
        };

      case 'interactive':
        console.log('Processing interactive response...');
        const interactiveInput: InteractiveAgentInput = {
          userInput: topic,
          latestContextSummary: this.learningState.sessionHistory
            .map(h => h.content)
            .join('\n')
        };
        const response = await this.callAgent(
          INTERACTIVE_AGENT_INSTRUCTIONS,
          interactiveInput
        ) as InteractiveAgentOutput;

        // Add to session history
        this.addToSessionHistory({
          type: 'explanation',
          content: response.response,
          timestamp: new Date().toISOString()
        });

        return {
          status: SafetyStatus.SAFE,
          explanation: response.response,
          subtopics: [],
          prerequisites: [],
          summary: ""
        };

      default:
        console.log('Starting new exploration...');
        // Check if input looks like an answer or uncertainty
        if (topic.toLowerCase().includes('not sure') || 
            topic.toLowerCase().includes('dont know') || 
            topic.toLowerCase().includes("don't know") ||
            topic.toLowerCase().includes('im unsure') ||
            topic.toLowerCase().includes("i'm unsure")) {
          console.log('Detected uncertainty, treating as answer evaluation...');
          return {
            status: SafetyStatus.SAFE,
            explanation: "That's okay! Let me help you understand this better. Would you like me to explain the topic again or give you a hint?",
            subtopics: ['Explain again', 'Give me a hint'],
            prerequisites: [],
            summary: ""
          };
        }

        // Default to exploration
        const explorationInput: ExplorationAgentInput = {
          userPrompt: topic,
          latestContextSummary: this.learningState.sessionHistory
            .map(h => h.content)
            .join('\n')
        };
        
        const exploration = await this.callAgent(
          EXPLORATION_AGENT_INSTRUCTIONS,
          explorationInput
        ) as ExplorationAgentOutput;

        // Reset and update learning state
        this.learningState = {
          ...this.initializeLearningState(),
          currentTopic: topic,
          learningPath: exploration.subtopics || []
        };

        // Add to session history
        this.addToSessionHistory({
          type: 'explanation',
          content: exploration.summary || '',
          timestamp: new Date().toISOString()
        });

        console.log('Pipeline complete');
        return exploration;
    }
  }

  // Gets explanation for a subtopic
  async getExplanation(subtopic: string): Promise<DeepDiveAgentOutput> {
    const input: DeepDiveAgentInput = {
      subtopic,
      broaderTopic: this.learningState.currentTopic,
      latestContextSummary: this.learningState.sessionHistory
        .map(h => h.content)
        .join('\n')
    };

    const explanation = await this.callAgent(
      DEEP_DIVE_AGENT_INSTRUCTIONS,
      input
    ) as DeepDiveAgentOutput;

    // Validate the explanation response
    if (!explanation || typeof explanation !== 'object') {
      throw new Error('Invalid explanation response format');
    }

    if (!explanation.breakdown || typeof explanation.breakdown !== 'string') {
      throw new Error('Invalid explanation content format');
    }

    // Format the content to include all components
    const formattedContent = `
# ${subtopic}

${explanation.breakdown}

${explanation.analogy ? '\n## Analogy\n' + explanation.analogy : ''}
${explanation.mermaidDiagram ? '\n## Diagram\n```mermaid\n' + explanation.mermaidDiagram + '\n```' : ''}
${explanation.codeExample ? '\n## Code Example\n```\n' + explanation.codeExample + '\n```' : ''}
    `.trim();

    // Update explanation content with formatted version
    explanation.breakdown = formattedContent;

    this.addToSessionHistory({
      type: 'explanation',
      content: explanation.breakdown,
      timestamp: new Date().toISOString()
    });

    return explanation;
  }

  // Generates a quiz question for given subtopic
  async getQuizQuestion(subtopic: string): Promise<QuestionAgentOutput> {
    console.log('\n=== Generating Quiz Question ===');
    console.log('Subtopic:', subtopic);
    
    try {
      const input: QuestionAgentInput = {
        subtopic,
        broaderTopic: this.learningState.currentTopic,
        latestContextSummary: this.learningState.sessionHistory
          .map(h => h.content)
          .join('\n')
      };

      const quiz = await this.callAgent(
        QUESTION_AGENT_INSTRUCTIONS,
        input
      ) as QuestionAgentOutput;

      // Validate quiz content with fallbacks
      if (!quiz || typeof quiz !== 'object') {
        console.log('Invalid quiz response, using fallback');
        return {
          question: `What have you learned about ${subtopic}?`,
          type: 'MCQ',
          options: [
            'I understand it well',
            'I need more explanation',
            'I have some questions',
            'Let\'s move on to the next topic'
          ],
          correctAnswer: 'I understand it well'
        };
      }

      // Ensure all required fields exist with fallbacks
      const validatedQuiz = {
        question: quiz.question || `What have you learned about ${subtopic}?`,
        type: quiz.type || 'MCQ',
        options: quiz.type === 'MCQ' ? (quiz.options || [
          'I understand it well',
          'I need more explanation',
          'I have some questions',
          'Let\'s move on to the next topic'
        ]) : undefined,
        correctAnswer: quiz.correctAnswer || (quiz.options?.[0] || 'I understand it well')
      };

      // Store the correct answer for feedback
      this.learningState.lastQuizAnswer = validatedQuiz.correctAnswer;

      console.log('Generated quiz:', validatedQuiz);
      return validatedQuiz;
    } catch (error) {
      console.error('Error generating quiz:', error);
      // Return a fallback question
      return {
        question: `What have you learned about ${subtopic}?`,
        type: 'MCQ',
        options: [
          'I understand it well',
          'I need more explanation',
          'I have some questions',
          'Let\'s move on to the next topic'
        ],
        correctAnswer: 'I understand it well'
      };
    }
  }

  // Provides feedback on user's answer
  async getFeedback(subtopic: string, userAnswer: string): Promise<AnswerEvalAgentOutput> {
    const input: AnswerEvalAgentInput = {
      subtopic,
      broaderTopic: this.learningState.currentTopic,
      questionAsked: this.learningState.sessionHistory
        .filter(h => h.type === 'quiz')
        .map(h => h.content)
        .slice(-1)[0],
      userQuestionAnswer: userAnswer,
      latestContextSummary: this.learningState.sessionHistory
        .map(h => h.content)
        .join('\n')
    };

    const feedback = await this.callAgent(
      ANSWER_EVAL_AGENT_INSTRUCTIONS,
      input
    ) as AnswerEvalAgentOutput;

    // Validate feedback response
    if (!feedback || typeof feedback !== 'object') {
      throw new Error('Invalid feedback response format');
    }

    if (typeof feedback.isCorrect !== 'boolean' || !feedback.feedback) {
      throw new Error('Invalid feedback content format');
    }

    this.addToSessionHistory({
      type: 'feedback',
      content: feedback.feedback,
      timestamp: new Date().toISOString()
    });

    // Update learning state
    if (feedback.isCorrect) {
      this.learningState.progress.masteredConcepts.push(subtopic);
    } else {
      this.learningState.progress.needsReview.push(subtopic);
    }

    return feedback;
  }

  // Gets session summary
  async getSessionSummary(): Promise<SummaryConsolidationAgentOutput> {
    const input: SummaryConsolidationAgentInput = {
      latestContextSummary: this.learningState.sessionHistory
        .map(h => h.content)
        .join('\n'),
      lastAgentInput: null,
      lastAgentOutput: null
    };

    return await this.callAgent(
      SUMMARY_CONSOLIDATION_AGENT_INSTRUCTIONS,
      input
    ) as SummaryConsolidationAgentOutput;
  }
}